internal/proposals/migration-hooks/migration-hooks-proposal.md

```diff
@@ -0,0 +1,233 @@
+- Only support single linear upgrade: A &#8594; B && B &#8594; C; NOT A &#8594; C. This is due to the fact that Kyma only supports single linear upgrades.
+- Mechanism should only trigger for one specific Kyma version, this could be: The Kyma version which wants to be installed on an empty cluster, Kyma version to upgrade to, or the Kyma version which should be uninstalled from the cluster.
+- This mechanism supports jobs for two different main use-cases: The __component based__ jobs and the __global/non-component based__ jobs
+  - __Component based__:
+    - Check if component is installed on cluster, or if it wants to be newly installed, and only trigger if yes
+    - It should be possible to trigger jobs before and after an deployment of a component
+  - __global / Component independent__:
+    - Always trigger logic when installing, upgrading, or uninstalling Kyma
+    - It should be possible to trigger jobs before and after the deployment, deletion of Kyma
+    - We call them `global` jobs to stick to the naming convention of our helm charts
+
+### Possible Solution
+
+To fulfill the requirements a new package, called `JobManager`, will be introduced, which is capable to register, manage, and trigger certain jobs to have a fully automated install/migration/deletion. This package has four (hash)maps to manage the main workload: Two for `pre`-jobs (deploy and deletion) and two for `post`-jobs (deploy and deletion). Key of the maps are the names of the component they belong to, and the value will be a slice of jobs.
+Furthermore, it has a `duration` var for benchmarking, and a `targetersion` var to know which jobs should be triggered at a certain deploy.
+
+Jobs will be implemented inside of the `JobManager` package in `go`-files, one for each component, using the specific `job`-interface. Then the implemented interface will be registered using `register(job)` in the same file. This function queues the jobs into a slice, this is due to the reason that we do not know what the value of the targetVersion is until now. 
+
+The JobManager will be used by the `deployment` package in the `deployment.go` and the `deletion.go` file. Inside of the `NewDeployment` or `NewDeletion` functions, the `SetKymaVersion` function of the JobManager will be called to set the targetVersion and thus to build the needed maps. Doing it in this way, we only have the requireed jobs registered in the maps and save some checks later on (&#8594; everything pre-calculated).Then at the hooks, during the deployment/deletion-phase, each hook just has to check if the key for the wanted component is present in the pre/post-map, if yes the jobs in the map will be trigged, if no nothing has to be done.
+
+To benchmark the jobs, a timer will be used in the pre and post job triggers.
+
+Retries for the jobs will not be handled by the JobManager. Retries should be implemented by the jobs themself. This is due to the reason to have a better flexability and a simple to manage interface.
```

Comment:
```suggestion
Retries for the jobs are not handled by the JobManager. Retries should be implemented by the jobs themselves, because it's more flexible and the interface is easy to manage.
```